src/core/gui/dialog/Dialog.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Xournal++
+ *
+ * Dialog base class
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <string>
+
+#include <gtk/gtk.h>  // for GtkWindow
+
+#include "gui/GladeGui.h"
+
+class DialogSPtr;
+class GladeSearchpath;
+
+class Dialog: public GladeGui {
+public:
+    template<class DialogClass, class... Args>
+    static DialogSPtr make(Args&&... args);
+
+protected:
+    Dialog(GladeSearchpath* gladeSearchPath, const std::string& glade, const std::string& mainWnd);
+    ~Dialog();
+
+private:
+    void show(GtkWindow* parent) override {}
+
+protected:
+    virtual void callCallback() const;
+    friend class DialogSPtr;
+};
+
+class DialogSPtr {
+public:
+    DialogSPtr(Dialog* dialog);
+    ~DialogSPtr();
+
+    void show(GtkWindow* parent);
+
+private:
+    static bool closeDialogCallback(Dialog* self);
+
+    Dialog* dialog = nullptr;
+};
```

Comment:
With the following, you can get rid of the `virtual void callCallback() const;` or just make all the functions + destructor in `Dialog` protected:

```suggestion
template <class DialogClass>
class DialogSPtr {
public:
    DialogSPtr(DialogClass* dialog);
    ~DialogSPtr();

    void show(GtkWindow* parent);

private:
    static bool closeDialogCallback(DialogClass* self);

    DialogClass* dialog = nullptr;
};
```